### PR TITLE
Add box model web component

### PR DESCRIPTION
**Pull request by williamw2**
# Pull Request

## 📖 Description

Adding a new box model web component to the fast-tooling package. This component renders a UI similar to the box model UI in the browser dev tools for modifying margin, border, padding, width and height css styles. When the component's "value" property is set to a valid CSS style string it will parse the values taking into account shorthand styles and order precedence to populate the UI. Then when any of the input fields are modified the component will attempt to construct the most efficient style string that represents the UI intent while preserving all other styles in the original string. The value property will then contain the new style string and a change event is emitted.

The component should correctly handle the following styles:
- margin
- margin-top
- margin-bottom
- margin-left
- margin-right
- padding
- padding-top
- padding-bottom
- padding-left
- padding-right
- width
- height
- border
- border-width
- border-top
- border-top-width
- border-bottom
- border-bottom-width
- border-left
- border-left-width
- border-right
- border-right-width

To test, pull down the branch, change to the /packages/fast-tooling folder and run "npm start". This will open the test site in a browser. Click the "CSS Box Model" button in the left menu.
Type a valid CSS string (i.e. "margin:10px;border:1px solid red;width:50%;") into the top text box and the UI should populate. Modify a value in one of the UI inputs and the bottom text box will update to the new CSS string.

### 🎫 Issues
Closes: #73 

### General

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.

### CODE REVIEW

File: change/@microsoft-fast-tooling-1e52c62c-7abe-471a-8c7b-3395a5265139.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add Box Model web component",
```
Comment (janechu) [inline change/@microsoft-fast-tooling-1e52c62c-7abe-471a-8c7b-3395a5265139.json:3]:
```suggestion
  "comment": "Added CSS box model web component",
```

File: packages/fast-tooling/src/web-components/color-picker/color-picker.spec.ts [hunk 2]
```diff
@@ -65,6 +76,8 @@ describe("ColorPicker", () => {
             wasChanged = true;
         });
 
+        console.log(element);
```
Comment (janechu) [inline packages/fast-tooling/src/web-components/color-picker/color-picker.spec.ts:79]: Still needed?

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.spec.ts [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+        marginTopInput.dispatchEvent(event);
+
+        await DOM.nextUpdate();
+
+        expect(wasChanged).to.equal(true);
+        expect(element.value).to.equal("margin-top:1px;");
+
+        await disconnect();
+    });
+    it("should correctly parse margin values", async () => {
```
Comment (janechu) [inline packages/fast-tooling/src/web-components/css-box-model/css-box-model.spec.ts:49]: nit: I would avoid the word "correctly" in the phrasing of the tests as it's redundant
```suggestion
    it("should parse margin values", async () => {
```

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+ */
+const cssLengthRegExp: RegExp = new RegExp(
+    /^auto$|^thin$|^medium$|^thick$|^[+-]?[0-9]+\.?([0-9]+)?(ch|rem|vh|vw|vmin|vmax|px|em|ex|%|in|cm|mm|pt|pc)?$/
+);
+
+enum valueType {
+    shortQuad = "shortQuad",
+    singleVal = "singleVal",
+    shortMixed = "shortMixed",
+    singleOverride = "singleOverride",
```
Comment (janechu) [inline packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts:26]: I'm going to take a stab at this, let me know if I interpreted it incorrectly:

```suggestion
    shorthandWithFourValues = "shorthand-with-four-values",
    oneValue = "one-value",
    shorthandWithMixedTypes = "shorthand-with-mixed-types",
    override = "override",
```

File: packages/fast-tooling/src/web-components/css-box-model/index.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { cssBoxModelTemplate as template } from "./css-box-model.template";
+import { cssBoxModelStyles as styles } from "./css-box-model.style";
+import { CSSBoxModel } from "./css-box-model";
+
+/**
+ * A web component used for updating CSS box model values.
+ *
+ * @alpha
+ * @remarks
+ * HTML Element: \<box-model\>
```
Comment (janechu) [inline packages/fast-tooling/src/web-components/css-box-model/index.ts:10]:
```suggestion
 * HTML Element: \<css-box-model\>
```

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+                this.resetValues();
+        }
+    }
+
+    /**
+     * Converts the values into the most efficient css string.
+     * @returns A string containing the the 1 to 4 parts of a margin, padding, border-width type style or an empty string
+     * if the values can not be correctly represented (one of the values is empty).
+     */
+    public getCSSQuadValue(): string {
```
Comment (janechu) [inline packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts:116]:
```suggestion
    public getCSSShorthandFourValues(): string {
```

File: change/@microsoft-fast-tooling-1e52c62c-7abe-471a-8c7b-3395a5265139.json (hunk 1 above)
Comment (williamw2) [reply change/@microsoft-fast-tooling-1e52c62c-7abe-471a-8c7b-3395a5265139.json]: Fixed

File: packages/fast-tooling/src/web-components/color-picker/color-picker.spec.ts (hunk 2 above)
Comment (williamw2) [reply packages/fast-tooling/src/web-components/color-picker/color-picker.spec.ts]: Nope. Removed.

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.spec.ts (hunk 3 above)
Comment (williamw2) [reply packages/fast-tooling/src/web-components/css-box-model/css-box-model.spec.ts]: Removed all instances.

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts (hunk 4 above)
Comment (williamw2) [reply packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts]: Changed

File: packages/fast-tooling/src/web-components/css-box-model/index.ts (hunk 5 above)
Comment (williamw2) [reply packages/fast-tooling/src/web-components/css-box-model/index.ts]: Fixed

File: packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts (hunk 6 above)
Comment (williamw2) [reply packages/fast-tooling/src/web-components/css-box-model/css-box-model.ts]: Changed

**Review by janechu — APPROVED** [review]

**Pull request merged**